### ISSUE TITLE
Support Provider<T> for org.gradle.process.JavaForkOptions#systemProperty value

### ISSUE THREAD

**Issue by vlsi** [gradle/gradle#12247]
### Expected Behavior

```kotlin
tasks.test {
    ...
    val libraryDir = developmentBinary.linkFile
        .map { it.asFile.parentFile.absolutePath }
    systemProperty("java.library.path", libraryDir)
}
```

It should lazily compute the property value only in case the task is considered for execution.

### Current Behavior

Currently, Gradle accepts `Object` value, however it calls `#toString()` to convert the value to string.
Of course, there's a workaround to supply an anonymous class that overrides `toString`

Here's a workaround:

```kotlin

/**
 * Gradle does not support [Provider] for [JavaForkOptions.systemProperty], so
 * we pass an object that overrides [Any.toString].
 */
fun Provider<String>.overrideToString() = object {
    override fun toString() = orNull ?: ""
}

...
tasks.test {
    dependsOn(developmentBinary.linkTask)
    val libraryDir = developmentBinary.linkFile
        .map { it.asFile.parentFile.absolutePath }
    systemProperty("java.library.path", libraryDir.overrideToString())
}
```

### Context

I'm building JNI-related project, and I want to pass the location of native library via system property, however it forces me to eagerly compute the location (e.g. even in case the task is not used).

I guess unwrapping `Providers` would not break backward compatibility much, as `Provider#toString` does not typically make much sense.

Gradle 6.1.1 if that matters.

**Comment by vlsi** [comment]
This is still needed. Would you please reopen the issue?

**Comment by vlsi** [comment]
Alternative workaround is to use
```kotlin
jvmArgumentProviders.add(...CommandLineArgumentProvider...)
```

**Comment by davidburstrom** [comment]
I came across the same missing functionality today, on Gradle 7.1.

**Comment by Vampire** [comment]
The work-around with `CommandLineArgumentProvider` is working fine and you can then also declare the input and output properties and implicit task dependencies properly.

But how do you make the `toString` work-around usable?
If I try to do it like mentioned, the build errors out with `Input property 'systemProperties' with value '{foo=bar}' cannot be serialized.`.

**Comment by vlsi** [comment]
>If I try to do it like mentioned, the build errors out with Input property 'systemProperties' with value '{foo=bar}' cannot be serialized..

I guess that depends on Gradle version you have, and the workaround seemed to work in 6.1.1.

Here's the full code for workaround, however it is no longer used: https://github.com/weisJ/darklaf/pull/33/files#diff-d74bc84ed5096fbbce671614a6ff36b789dea317b6b233758c97f35b6862142eR53-R68

**Comment by davidburstrom** [comment]
Could it be that your lambda is a closure? If it binds an outer context, e.g. the `Project` object, it cannot be serialized.

**Comment by Vampire** [comment]
Whatever I do, I get `Caused by: java.io.NotSerializableException: Build_gradle$overrideToString$1` or similar.
Making the anonymous object serializable also does not help as you then get `Caused by: java.io.NotSerializableException: org.gradle.api.internal.provider.DefaultProperty` or similar unless you happen to have a serializable provider.

The code that @vlsi linked to just seems to work because it is not used.
If I check out that PR branch of that project, and do `gradlew darklaf-windows:test -x compileDebugX86-64Cpp` (the exclude because I have no cpp toolchain, but it doesn't matter for this), then I get `Task :darklaf-windows:test NO-SOURCE`.
If I then do `mkdir -p windows/src/test/java` followed by `echo 'public class Foo {}' >windows/src/test/java/Foo.java` and then again `gradlew darklaf-windows:test -x compileDebugX86-64Cpp`, I get the same "cannot be serialized" problem with `Caused by: java.io.NotSerializableException: Jni_library_gradle$overrideToString$`.

**Comment by melix** [comment]
I agree this is important: I've faced this issue 4 times _just this week_ on different projects: not being able to pass a `Provider<>` as a `systemProperty` of a `Test` task is particularly harmful.

**Comment by melix** [comment]
I faced this problem again, not for tests but for `JavaExec`: see https://github.com/micronaut-projects/micronaut-gradle-plugin/pull/294 and even more horrific https://github.com/micronaut-projects/micronaut-gradle-plugin/commit/e93fef23fe06657f734779ac5ab3bef947f5608e#diff-1abcd1c02a218798ff74713636d9db7f6401f1ac5620490b1492d107d9482ab6R143

**Comment by eskatos** [comment]
This should be fixed together with environment variables that have the same problem.

**Comment by lptr** [comment]
We are planning to address this in 9.0 by converting `JavaForkOptions.systemProperty` to `MapProperty` which should take `Provider` values.

**Comment by rkrisztian** [comment]
Thanks for addressing it, I have a similar problem for which I currently do this:

```groovy
testing {
	suites {
		integrationTest(JvmTestSuite) {
			targets.all {
				testTask.configure {
					systemProperty 'java.io.tmpdir', "${-> layout.buildDirectory.dir('tmp').get()}"
				}
			}
		}
	}
}
```

**Comment by pantherdd** [comment]
My current workaround for using lazily computes values as a `systemProperty` is this (Kotlin DSL):
```kts
class LazyString(private val source: Lazy<String>) : java.io.Serializable {
    constructor(source: () -> String) : this(lazy(source))
    constructor(source: Provider<String>) : this(source::get)
    override fun toString() = source.value
}
```
Then you can do stuff like:
```kts
val now: Provider<Instant> = provider { Instant.now() }
tasks.test {
    systemProperty("bySupplier", LazyString { now.get().toString() })
    systemProperty("byProvider", LazyString(now.map(Instant::toString)))
}
```
Since it uses `lazy`, the value is going to be memoized once calculated, so it's safe in that sense. Of course one could even make it accept sources that return `Object`, and make the `toString()` call within `LazyString`, but I liked the type safety of requiring a `String` explicitly.

**Comment by ljacomet** [comment]
In Kotlin, one can directly use `by lazy` for this:
```
val myLazyValue: String by lazy {
    // Compute/derive the value
}
systemProperty("name", myLazyValue)
```

This is what I end up doing in the following PR to improve CC hits on the Gradle build itself:
* #32487

**Comment by sdeleuze** [comment]
I am a bit confused, it looks like a popular ask, it was planned for Gradle 9.0, but it looks like it has been cancelled without any context shared. Any chance we could have a status update?

**Comment by Vampire** [comment]
The big providerization was shifted to Gradle 10: https://github.com/gradle/build-tool-roadmap/issues/28#issuecomment-2761751557
